### PR TITLE
feature(crypto): Add support for master key local pinning

### PR DESCRIPTION
**Pull request by BillCarsonFr**
part of invisible crypto, follow up of https://github.com/matrix-org/matrix-rust-sdk/pull/3607
Alternative fix for  https://github.com/matrix-org/matrix-rust-sdk/issues/3564
Alternative PR https://github.com/matrix-org/matrix-rust-sdk/pull/3610

Add the capability to locally pin a public MSK for a `ReadOnlyUserIdentity`. 
The first time an identity is seen for a user, the msk is pinned. Pin violation will be reported when an identity is rotated.

This PR only adds support for pinning, support for serialization/migration, persistance. 
This could be used later by other PRs to report specific errors or show pinning violation to users.

Note about verification and pinning.
As part of this PR, if a new identity is detected it will still be seen as a pinning violation even if the new identity is signed by our usk. But the `UserIdentity::has_identity_mismatch()` will be ok. There is a pinning violation, but the new identity is verified and verification has priority. 
That is to say that there is no auto-pinning if verified so far. To be discussed if we want it later

- [ ] Public API changes documented in changelogs (optional)

Signed-off-by:

### CODE REVIEW

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/identities/user.rs:405]: It might be cleaner to write custom `Serialize` and `Deserialize` implementations, but I'm not sure.

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/identities/user.rs:416]: Worth noting the presence of this field means that the documentation on this class ("other users will
only contain a master key and a self signing key"), as well as its name, are no longer correct. We should at least fix the doc; fixing the name would be good too but might be a bit invasive (plus, we'd need to think of good names).

**Review by richvdh — COMMENTED** [review]
This looks broadly sensible to me, though I'd defer to @poljar on this.

There are a bunch of documentation nits I'd like to see cleaned up, but let's agree on the basic structure first

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (BillCarsonFr) [reply crates/matrix-sdk-crypto/src/identities/user.rs:405]: I tried and for it to work properly I had to put a new field `version: Option<u8>` in the `ReadOnlyUserIdentity` model. 
I tried then with the helper pattern seen in other part of the code and it looks a bit better I think as it delegates serialization to a specific object and just needs standard `From` impl.

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (poljar) [inline crates/matrix-sdk-crypto/src/identities/user.rs:575]: The first line should be a short explanation of the method kept under some 80-100 characters, then we can expand on the first line in the next paragraph.

Could you try to reword this to conform this style?

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (poljar) [inline crates/matrix-sdk-crypto/src/identities/user.rs:316]: Similar to the `has_pin_violation()` method, the first line should be something akin to "Has the identity changed since we last pinned it?"

I think this could also could be perhaps better named, identity mismatch makes me think of the identity not being what we expect. I guess you could argue that this is exactly the case here, but perhaps we could name it to be something along the lines that the identity has changed without our approval instead. Feel free to ignore this part if you disagree.

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (poljar) [inline crates/matrix-sdk-crypto/src/identities/user.rs:1217]:
```suggestion
    async fn resolve_identity_mismatch_with_verification() {
```

File: testing/matrix-sdk-test/src/test_json/keys_query_sets.rs
Comment (poljar) [inline testing/matrix-sdk-test/src/test_json/keys_query_sets.rs:637]: But? But what?!? This cliff-hanger produces anxiety.

File: testing/matrix-sdk-test/src/test_json/keys_query_sets.rs
Comment (poljar) [inline testing/matrix-sdk-test/src/test_json/keys_query_sets.rs:619]: More butts.

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (poljar) [inline crates/matrix-sdk-crypto/src/identities/user.rs:599]: Perhaps we can expand this a bit, to explain the why. I know the why right now but another reader might not know it.

File: crates/matrix-sdk-crypto/src/identities/manager.rs
Comment (poljar) [inline crates/matrix-sdk-crypto/src/identities/manager.rs:1968]:
```suggestion
        // We receive updated keys for that user, with no identity anymore
```

**Review by poljar — CHANGES_REQUESTED** [review]
I think the approach looks good, I left a bunch of nits though. I'll let @dkasak double check things.

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (BillCarsonFr) [reply crates/matrix-sdk-crypto/src/identities/user.rs:319]: I am not sure what would be the best name for that. I took identity mismatch to use the same term that is used in this [doc](https://github.com/element-hq/element-meta/blob/develop/docs/crypto/trust_v2.md#behavior-for-identity-mismatch).

Maybe @dkasak have an idea? Maybe `is_identity_trustworthy`? or more based on the use of this method `identity_needs_user_approval` / `validation` / `review` ?

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (poljar) [inline crates/matrix-sdk-crypto/src/identities/user.rs:618]: This won't produce a link since it isn't a doc comment, at least it doesn't for me, for you it does?
```suggestion
        // (see `has_pin_violation()`).
```

**Review by poljar — APPROVED** [review]
Left a tiny nit, but I think that this is fine.

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (dkasak) [reply crates/matrix-sdk-crypto/src/identities/user.rs:323]: I would avoid the `mismatch` and `trustworthy` wording. `identity_needs_user_approval` sounds good to me.

File: crates/matrix-sdk-crypto/src/identities/manager.rs
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/identities/manager.rs:1926]:
```suggestion
        // We should now have an identity for the user but no pin violation
        // (pinned master key is the current one)
```

File: crates/matrix-sdk-crypto/src/identities/manager.rs
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/identities/manager.rs:1975]:
```suggestion
        // We receive updated keys for that user, with no identity anymore.
        // Notice that there is no server API to delete identity, but we want to
        // test here that a home server cannot clear the identity and
        // subsequently serve a new one which would get automatically approved.
```

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/identities/user.rs:300]:
```suggestion
    /// Pin the current identity (public part of the master signing key).
```

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/identities/user.rs:313]:
```suggestion
    /// Did the identity change after an initial observation in a way that
    /// requires approval from the user?
```

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/identities/user.rs:318]:
```suggestion
    /// A user identity needs approval if it changed after the crypto machine
    /// has already observed ("pinned") a different identity for that user *and*
    /// it is not an explicitly verified identity (using for example interactive
    /// verification).
    ///
    /// Such a change is to be considered a pinning violation which the
    /// application should report to the local user, and can be resolved by:
    ///
```

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/identities/user.rs:321]:
```suggestion
    /// - Or by updating the pin to the new identity
```

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/identities/user.rs:420]:
```suggestion
/// This struct also contains the currently pinned user identity (master signing
/// key) for that user.
///
/// The first time a cryptographic user identity is seen for a given user, it
/// will be associated with that user ("pinned"). Future interactions
/// will expect this identity to stay the same, to avoid MITM attacks from the
/// homeserver.
///
/// The user can explicitly pin the new identity to allow for legitimate
/// identity changes (for example, in case of key material or device loss).
```

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/identities/user.rs:1251]:
```suggestion
        // Simulate an identity change
```

File: crates/matrix-sdk-crypto/src/identities/manager.rs
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/identities/manager.rs:2021]:
```suggestion
        // Resolve the mismatch by pinning the new identity
```

File: crates/matrix-sdk-crypto/src/identities/user.rs
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/identities/user.rs:432]:
```suggestion
/// versioning and migration.
///
/// Version v1 is adding support for identity pinning (`pinned_master_key`), as
```

**Review by dkasak — APPROVED** [review]
I've left a bunch of wording suggestions and would advise changing the `has_identity_mismatch` method name to `identity_needs_user_approval` as discussed, but otherwise this seems good to me, so leaving an `Approve` for when suggestions are handled.

**Pull request merged**
